File: problems/0133-clone-graph/typescript/solution.ts
// Clone Graph — https://leetcode.com/problems/clone-graph/
class GraphNode {
  val: number;
  neighbors: GraphNode[] = [];
  constructor(val: number) {
    this.val = val;
  }
}

function cloneGraph(node: GraphNode | null): GraphNode | null {
  if (node === null) return null;
  const clones = new Map<GraphNode, GraphNode>();
  const dfs = (n: GraphNode): GraphNode => {
    const existing = clones.get(n);
    if (existing !== undefined) return existing;
    const copy = new GraphNode(n.val);
    clones.set(n, copy);
    for (const nb of n.neighbors) copy.neighbors.push(dfs(nb));
    return copy;
  };
  return dfs(node);
}

export function cloneGraphArr(adjList: number[][]): number[][] {
  const n = adjList.length;
  if (n === 0) return [];
  const nodes = Array.from({ length: n }, (_, i) => new GraphNode(i + 1));
  adjList.forEach((neighbors, i) => {
    nodes[i].neighbors = neighbors.map((v) => nodes[v - 1]);
  });
  const cloned = cloneGraph(nodes[0]);
  const result: number[][] = Array.from({ length: n }, () => []);
  if (cloned === null) return result;
  const visited = new Set<number>([cloned.val]);
  const queue: GraphNode[] = [cloned];
  while (queue.length > 0) {
    const cur = queue.shift();
    if (cur === undefined) break;
    result[cur.val - 1] = cur.neighbors.map((nb) => nb.val);
    for (const nb of cur.neighbors) {
      if (!visited.has(nb.val)) {
        visited.add(nb.val);
        queue.push(nb);
      }
    }
  }
  return result;
}

File: problems/0133-clone-graph/go/solution.go
package solution

// Clone Graph — https://leetcode.com/problems/clone-graph/
type GraphNode struct {
	Val       int
	Neighbors []*GraphNode
}

func cloneGraph(node *GraphNode) *GraphNode {
	if node == nil {
		return nil
	}
	clones := map[*GraphNode]*GraphNode{}
	var dfs func(n *GraphNode) *GraphNode
	dfs = func(n *GraphNode) *GraphNode {
		if c, ok := clones[n]; ok {
			return c
		}
		copy := &GraphNode{Val: n.Val}
		clones[n] = copy
		for _, nb := range n.Neighbors {
			copy.Neighbors = append(copy.Neighbors, dfs(nb))
		}
		return copy
	}
	return dfs(node)
}

func cloneGraphArr(adjList [][]int) [][]int {
	n := len(adjList)
	if n == 0 {
		return [][]int{}
	}
	nodes := make([]*GraphNode, n)
	for i := 0; i < n; i++ {
		nodes[i] = &GraphNode{Val: i + 1}
	}
	for i, neighbors := range adjList {
		for _, v := range neighbors {
			nodes[i].Neighbors = append(nodes[i].Neighbors, nodes[v-1])
		}
	}
	cloned := cloneGraph(nodes[0])
	result := make([][]int, n)
	for i := range result {
		result[i] = []int{}
	}
	visited := map[int]bool{cloned.Val: true}
	queue := []*GraphNode{cloned}
	for len(queue) > 0 {
		cur := queue[0]
		queue = queue[1:]
		vals := []int{}
		for _, nb := range cur.Neighbors {
			vals = append(vals, nb.Val)
			if !visited[nb.Val] {
				visited[nb.Val] = true
				queue = append(queue, nb)
			}
		}
		result[cur.Val-1] = vals
	}
	return result
}
